libs/core/schedulers/include/hpx/schedulers/thread_queue.hpp

```diff
@@ -249,21 +255,14 @@ namespace hpx { namespace threads { namespace policies {
                 // Decrement only after thread_map_count_ has been incremented
                 --addfrom->new_tasks_count_.data_;
 
-                // only insert the thread into the work-items queue if it is in
-                // pending state
-                if (schedule_now)
-                {
-                    // pushing the new thread into the pending queue of the
-                    // specified thread_queue
-                    ++added;
-                    schedule_thread(get_thread_id_data(thrd));
-                }
+                // insert the thread into the work-items queue assuming it is
+                // in pending state, thread would go out of scope otherwise
+                HPX_ASSERT(schedule_now);
```

Comment:
This assert seems odd to me, but I'm probably missing the bigger context. Why can we assume that `schedule_now` is always true here? The comment above makes it sound like it can still be false sometimes, but the assert doesn't match that.

Reply:
In general, the scheduling system now keeps threads alive if they are `pending` (sitting in a queue), `active`/running (the scheduling loop holds them), are `suspended` (there is a reference count sitting on the thread's stack created before the actual suspension), or explicitly by the user. 

Therefore, staged tasks must have their initial state set to `pending`. They are converted to real threads without user intervention by the scheduler, thus the only way to keep them alive (keep the reference count up) is to schedule them (add them to a queue) once they have been created.

This assert should be always true because of this: https://github.com/STEllAR-GROUP/hpx/pull/5441/files/c2965e6bfee08b87582f46c6b0521b15f1cd6738#diff-34057d72d801cb0ee81de2f396edd2f1de7bd0d6309445a06df1d1e58630de8eR733. I should however turn the second assert into an exception to enforce the first.

Reply:
Do we then even need the `staged` state anymore?

Reply:
The staged state represents a thread to be delay-created only once the scheduler runs out of work,  so this question is an orthogonal one. The new limitation really just requires for staged threads to be scheduled right away once they are created eventually. I think this is what I would expect anyways. What use is a staged task if it should not be run once turned into a 'real' thread?

Reply:
Fully agreed. The `run_now` flag represents whether or not a thread should start as `staged` or be scheduled immediately, not the `staged` state. As far as I can tell, and this seems to have been true before this PR as well, the only places we use `thread_schedule_state::staged` is to query the sizes of the staged tasks queues. Neither `thread_init_data` nor `thread_data` should ever be in the `staged` state (there's an `initial_state_` in `thread_init_data` but that's for when the thread actually is scheduled).

With that in mind should we also add an assertion that `initial_state_` is never `staged` to the `thread_init_data` constructors: https://github.com/STEllAR-GROUP/hpx/blob/fa6e4df704bf5f81ddc0f4e09716bd11ff8724c8/libs/core/threading_base/include/hpx/threading_base/thread_init_data.hpp#L29-L133? In addition to the assertions that you've already added and proposed, that is.

Reply:
I agree that the initial state should not be `staged`, ever. It's an internal state users shouldn't have to worry about. I will add an error check for this.